src/labs.txt

```diff
@@ -2745,6 +2745,200 @@ Output:
 =hist
 EOF
 
+----------------------------------------------------------------------
+h1. Finding Commits That Introduced a Change
+
+h2. Goals
+
+* Learn how to use @git bisect@ to find a commit that introduced a change.
+
+p. It is sometimes necessary to find a commit that introduced a specific
+change.  Commonly you will be searching for a bug, but sometimes the change
+you're looking for is a feature. Git allows you to efficiently search for a
+commit as long as you can describe your search criteria in terms of a command
+to run.
+
+p. In this lab we'll be searching for the commit that allowed users to specify
+their name. While you could find it manually in this project, when there are
+hundreds or thousands of commits this becomes impractical. @git bisect@ will
+make this task trivial.
+
+h2. The Test Command
+
+p. The first thing we need is a test command. @git bisect@ works by checking
+out different commits and seeing if the test command succeeds or fails. For
+what we're searching for we'll write a small bash script.
+
+File: bisect.sh
+#!/usr/bin/env bash
+
+if [ -e lib/hello.rb ]
+then
+    if ruby -Ilib lib/hello.rb BisectString | grep BisectString
+    then
+        echo "New"
+        exit 1
+    else
+        echo "Old"
+        exit 0
+    fi
+elif [ -e hello.rb ]
+then
+    if ruby hello.rb BisectString | grep BisectString
+    then
+        echo "New"
+        exit 1
+    else
+        echo "Old"
+        exit 0
+    fi
+fi
+EOF
+
+p. Our repository contains two locations of the @hello.rb@ file at different
+times and our test script needs to take this into account. It first checks if
+the file is in the @lib/@ directory, and if it is it executes it. If the file
+is not there then it checks if it is in our original location and executes it
+there. In both cases, we are searching the output of the command for our input
+string to determine if the program exhibits the behavior we are searching for,
+and finally printing whether it does or not. You can test the script now.
+
+Execute:
+chmod +x bisect.sh
+./bisect.sh
+=scripttest
+
+Output:
+=scripttest
+EOF
+
+Freeze
+
+h2. Using @git bisect@
+
+p. Once you have your test command using @git bisect@ is fairly
+straightforward. You need a known location that doesn't exhibit the behavior
+and a known location that does. @git bisect@ calls these "bad" and "good" or
+"new" and "old" commits respectively. For our purposes, the new commit is our
+@HEAD@, and the old commit is the initial commit. Let's get started.
+
+Set: hash=hash_for("First Commit")
+Execute:
+git bisect start
+=bisectstart1
+git bisect old <hash>
+=bisectstart2
+git bisect new HEAD
```

Comment:
New/old fits the "when did this change happen" model better than good/bad. I actually didn't know about it until I was looking at the manual for bisect while doing this lab. Given the small scale of this project there aren't really any bugs and I didn't want to clutter an already complicated lab by adding an artificial bug so I went with the slightly less contrived approach of finding when a change happened.

Reply:
I think that good/bad is the more "typical" terminology people will be using though (and do use day to day).  As you said as well, you didnt even know about new/old until researching this lab.

Bisect is (most?) often used to "hunt down where a bug was introduced", and as such framing the lab in the scope of what the normal/average usage of the command makes a bit more sense to me.

I do agree that it could seem weird to introduce an artificial bug into the lab, but I'd have to think about if I think that is a more or less contrived example to use.  I guess it's probably different sides of the same coin as we have to craft some sort of example use-case to allow the lab to exist.

Reply:
So my biggest concern with crafting a bug is for the bisect to demonstrate its value the bug would have to be somewhere relatively deep in the history. Starting the lab by saying, "Write this bug and commit it" would just leave bisect to point at the commit they just created and I felt like that wouldn't demonstrate the point very well. Going with `old/new` instead of `good/bad` felt like it preserved the value of the bisect better to me.

Reply:
Let's stick with "good"/"bad".  People can learn obscure options from the man pages.